### PR TITLE
Us574

### PR DESCRIPTION
**Pull request by kescob0825**

### CODE REVIEW

File: Assets/Scripts/Managers/TurnFlow/TurnActionType.cs
Comment (hdathert) [inline Assets/Scripts/Managers/TurnFlow/TurnActionType.cs:1]: This doesn't capture all of the turn actions we can take as a player- we're missing 'sell upgrade', 'mortgage property', and 'unmortgage property'. But based on how this enum is used, it seems like we could change `UpgradeProperty` to something like `ModifyProperty`, which is also allowed to happen at any point in the turn. This captures all of the actions and doesn't require modifying the underlying flow coordinator and is much more semantically clear on what action is being taken, or more, what buckets those belong to.

File: Assets/Scripts/Managers/PlayerControllers/HumanPlayerController.cs [hunk 1]
```diff
@@ -71,16 +74,49 @@ public override void Unsubscribe()
...
-            if (!isMyTurn) return;
-
-            uiActivationEventChannel.RaiseEvent(
-                new UIActivationEvent(
-                    UIType.PropertyPurchase,
-                    new PurchaseActivationContext(
-                        pore.requestedSpace,
-                        pore.cost,
-                        controlledPlayer.CanAfford(pore.requestedSpace.buyPrice)))); 
+            // Note: Follow this pattern for any event that requires player input.
```
Comment (hdathert) [inline Assets/Scripts/Managers/PlayerControllers/HumanPlayerController.cs:86]: Why are we not doing internal turn checking anymore? Removing this check means that every player for every action is going to send events to the turn manager, which seems really wasteful, when we can just make the 'isMyTurn' check and guard against that.

We have the benefit of that early exit- we should use it. Having additional validation in the turn coordinator makes sense as well, but it seems like removing that check is over-correcting.

Comment (hdathert) [reply Assets/Scripts/Managers/PlayerControllers/HumanPlayerController.cs:86]: Just for clarity, I mean that when *any* player lands on an unowned space, the PurchaseOwnableRequestEvent fires, and since *every* player is subscribed, this method will run on *every* player. Having an early exit means we aren't sending 2-3 extra events every turn, potentially causing more issues.

File: Assets/Scripts/Managers/PlayerControllers/HumanPlayerController.cs [hunk 2]
```diff
@@ -71,16 +74,49 @@ public override void Unsubscribe()
...
+                {
+                    Logger.Debug("HumanPlayerController.HandlePurchaseOwnableEvent",
+                        "Purchase UI blocked by TurnFlow.",
+                        LogCategory.UI);
+                });
+        }
 
+        private void HandleUpgradeEvent(/*TODO: add event data here*/)
+        {
+            // Note: Follow this pattern for any event that requires player input.
```
Comment (hdathert) [inline Assets/Scripts/Managers/PlayerControllers/HumanPlayerController.cs:109]: This method doesn't need checking in the same way that Purchase does, because there is a handling method for UI Action events that routes UI actions to the relevant PlayerController methods like this one- so this is all fine.

**Review by hdathert — CHANGES_REQUESTED** [review]
Needs a few minor adjustments, I like this pattern though.

**Review by hdathert — APPROVED** [review]
# Code Review Checklist

**Completed for**: US #574 and Tasks #607 to #610  
**Code Review Completed By**: @hdathert 

Use this checklist to review new code changes before merging.

---

## Implementation
- [x] Does this code change accomplish what it is supposed to do?
- [x] Can the solution be simplified?
- [x] This change does **not** add unwanted compile-time or run-time dependencies.
- [x] The code is at the right abstraction level.
- [x] The code is modular enough.
- [x] The solution is adequate for maintainability, readability, performance, and security.
- [x] Similar functionality does **not** already exist in the codebase (or reuse is not appropriate).
- [x] No additional best practices or design patterns would substantially improve this code.

---

## Logic Errors and Bugs
- [x] No use cases where the code behaves incorrectly.
- [x] No inputs or external events that can break the code.
- [x] Code agrees with specifications.
- [x] No off-by-one errors.
- [x] No magic numbers (or they are well-named constants).
- [x] Proper use of global/large-scope variables (minimized and justified).

---

## Error Handling and Logging
- [x] Error handling is implemented correctly.
- [x] Error messages are user-friendly (where applicable).
- [x] Sufficient, meaningful log events exist for debugging (no sensitive data in logs).

---

## Dependencies
- [x] Documentation/configuration/README updates were made as needed.
- [x] No negative impact on other parts of the system or backward compatibility.

---

## Security and Data Privacy
- [x] No new security vulnerabilities introduced.
- [x] All user input is validated, sanitized, and escaped (XSS/SQL injection prevented).
- [x] Data from external APIs/libraries is handled safely and checked as needed.

---

## Performance
- [x] No significant performance regressions.
- [x] No obvious opportunities for significant performance gains left unaddressed.

---

## Ethics and Morality
- [x] Appropriate measures exist to prevent or report harassment/abuse (if applicable).

---

## Testing and Testability
- [x] Code is testable.
- [x] Automated tests have been added or updated.
- [x] Existing tests sufficiently cover the change (unit/integration/system).
- [x] Edge cases and boundary inputs are tested.

---

## Readability
- [x] Code is easy to understand.
- [x] Methods/functions are appropriately small and focused.
- [x] Names (functions/classes/variables) are clear and descriptive.
- [x] Code is in the correct file/folder/package.
- [x] Control flow is intuitive; data flow is clear.
- [x] Comments are clear, relevant, and up to date.
- [x] No commented-out or redundant code remains.
- [x] No overly dense one-liners; no overly long methods.

---

**Use this before submitting a pull request.**  
It helps ensure code quality, maintainability, and compliance with best practices.

*Based on awesome code review guides and MIT’s code review recommendations.*

## There is an issue in HumanPlayerController that came from another PR that was merged, but thats been fixed in US583, since I thought it came from me.

**Comment by tsaldrid** [comment]
# Code Review Checklist

**Completed for**: US #574 and Tasks #607 to #610
**Code Review Completed By**: @tsaldrid

Use this checklist to review new code changes before merging.

---

## Implementation
- [x] Does this code change accomplish what it is supposed to do?
- [x] Can the solution be simplified?
- [X] This change does **not** add unwanted compile-time or run-time dependencies.
- [x] The code is at the right abstraction level.
- [x] The code is modular enough.
- [x] The solution is adequate for maintainability, readability, performance, and security.
- [x] Similar functionality does **not** already exist in the codebase (or reuse is not appropriate).
- [x] No additional best practices or design patterns would substantially improve this code.

---

## Logic Errors and Bugs
- [x] No use cases where the code behaves incorrectly.
- [x] No inputs or external events that can break the code.
- [x] Code agrees with specifications.
- [x] No off-by-one errors.a
- [x] No magic numbers (or they are well-named constants).
- [x] Proper use of global/large-scope variables (minimized and justified).

---

## Error Handling and Logging
- [x] Error handling is implemented correctly.
- [x] Error messages are user-friendly (where applicable).
- [x] Sufficient, meaningful log events exist for debugging (no sensitive data in logs).

---

## Dependencies
- [x] Documentation/configuration/README updates were made as needed.
- [x] No negative impact on other parts of the system or backward compatibility.

---

## Security and Data Privacy
- [x] No new security vulnerabilities introduced.
- [x] All user input is validated, sanitized, and escaped (XSS/SQL injection prevented).
- [x] Data from external APIs/libraries is handled safely and checked as needed.

---

## Performance
- [x] No significant performance regressions.
- [x] No obvious opportunities for significant performance gains left unaddressed.

---

## Ethics and Morality
- [x] Appropriate measures exist to prevent or report harassment/abuse (if applicable).

---

## Testing and Testability
- [x] Code is testable.
- [x] Automated tests have been added or updated.
- [x] Existing tests sufficiently cover the change (unit/integration/system).
- [x] Edge cases and boundary inputs are tested.

---

## Readability
- [x] Code is easy to understand.
- [x] Methods/functions are appropriately small and focused.
- [x] Names (functions/classes/variables) are clear and descriptive.
- [x] Code is in the correct file/folder/package.
- [x] Control flow is intuitive; data flow is clear.
- [x] Comments are clear, relevant, and up to date.
- [x] No commented-out or redundant code remains.
- [x] No overly dense one-liners; no overly long methods.

---

**Use this before submitting a pull request.**  
It helps ensure code quality, maintainability, and compliance with best practices.

*Based on awesome code review guides and MIT’s code review recommendations.*
